### PR TITLE
Add new step to deploy wizard that configures parameter overrides.

### PR DESCRIPTION
**Pull request by mpiroc**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

Add a new step to the deploy wizard that reads parameters from the user's SAM template and overrides from `templates.json`, and prompts the user to configure `templates.json` if appropriate.

## Motivation and Context

Parameter overrides in `templates.json` are not easily discoverable on their own. We want to direct users to this file, but only when necessary.

## Related Issue(s)

#331 
#380

## Testing

* Added new tests
* Ran all new and existing tests

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lambda/wizards/samDeployWizard.ts [hunk 1]
```diff
@@ -30,13 +33,47 @@ export interface SamDeployWizardContext {
...
      */
     promptUserForSamTemplate(initialValue?: vscode.Uri): Promise<vscode.Uri | undefined>
 
+    /**
+     * Prompts the user to configure parameter overrides, then either pre-fills and opens
+     * `templates.json`, or returns true.
+     *
+     * @param options.templateUri The URL of the SAM template to inspect.
+     * @param options.missingParameters The names of required parameters that are not yet overridden.
+     * @returns A value indicating whether the wizard should procede. `false` if `missingParameters` was
```
Comment (awschristou) [inline src/lambda/wizards/samDeployWizard.ts]: procede -> proceed

File: src/lambda/wizards/samDeployWizard.ts [hunk 2]
```diff
@@ -100,6 +151,89 @@ class DefaultSamDeployWizardContext implements SamDeployWizardContext {
...
+                options: {
+                    title: prompt
+                },
+                items: [
+                    { label: responseYes },
+                    { label: responseNo }
+                ]
+            })
+            const response = getSingleResponse(await picker.promptUser({ picker: quickPick }))
+            if (response === responseYes) {
```
Comment (awschristou) [inline src/lambda/wizards/samDeployWizard.ts]: should this be === responseNo to return true? I understand returning true means "carry on with the deployment" but yes means "take me to the editor".

File: src/lambda/wizards/samDeployWizard.ts [hunk 3]
```diff
@@ -264,12 +405,61 @@ export class SamDeployWizard extends MultiStepWizard<SamDeployWizardResponse> {
...
+        this.response.parameterOverrides = overriddenParameters
+
+        return this.REGION
     }
 
     private readonly REGION: WizardStep = async () => {
         this.response.region = await this.context.promptUserForRegion(this.regionProvider, this.response.region)
 
+        // The PARAMETER_OVERRIDES step is part of the TEMPLATE step from the user's perspective,
+        // so we go back to the TEMPLATE step instead of PARAMETER_OVERRIDES.
```
Comment (awschristou) [inline src/lambda/wizards/samDeployWizard.ts]: ✔

File: src/lambda/wizards/samDeployWizard.ts [hunk 4]
```diff
@@ -30,13 +33,47 @@ export interface SamDeployWizardContext {
...
+    /**
+     * Prompts the user to configure parameter overrides, then either pre-fills and opens
+     * `templates.json`, or returns true.
+     *
+     * @param options.templateUri The URL of the SAM template to inspect.
+     * @param options.missingParameters The names of required parameters that are not yet overridden.
+     * @returns A value indicating whether the wizard should procede. `false` if `missingParameters` was
+     *          non-empty, or if it was empty and the user opted to configure overrides instead of continuing.
+     */
+    configureParameterOverrides(
```
Comment (awschristou) [inline src/lambda/wizards/samDeployWizard.ts]: the name `configureParameterOverrides` and the return value meanings are not intuitive. I had a hard time interpreting the calling code around these as I went through the PR.

Consider something like `stopToEditParameterOverrides` or something like that (might flip the meaning of return values)

File: src/lambda/wizards/samDeployWizard.ts (hunk 2 above)
Comment (mpiroc) [reply src/lambda/wizards/samDeployWizard.ts]: ✔️

File: src/test/lambda/wizards/samDeployWizard.test.ts [hunk 5]
```diff
@@ -187,6 +208,219 @@ describe('SamDeployWizard', async () => {
         })
     })
 
+    describe('PARAMETER_OVERRIDES', async () => {
+        function makeFakeRegionProvider(): RegionProvider {
```
Comment (bryceitoc9) [inline src/test/lambda/wizards/samDeployWizard.test.ts]: Why not use `MockRegionProvider`?

Comment (mpiroc) [reply src/test/lambda/wizards/samDeployWizard.test.ts]: Good catch, I forgot to replace this when I added `MockRegionProvider`. Will do.

**Pull request merged**